### PR TITLE
Global styles menu: Avoid visible labels and accessible names mismatch

### LINKED ISSUE #65112: Global styles menu: avoid unnecessary mismatch between visible label and accessible name
### Description

See related discussion on https://github.com/WordPress/gutenberg/issues/22332

Cc @joedolson @alexstine 

For the highest level of accessibility, the _visible label_ of a control should determine the control's accessible name.

Sometimes, with good intent, both Core and Gutenberg use the `aria-label` attribute to provide more context but by doing so a mismatch between the visible label and the actual accessible name occurs.

While the expanded info in the `aria-label` may be beneficial for blind screen reader users, it introduces potential barriers for other users. For example:
- Sighted screen reader users will see the visible label but the screen reader will announce the invisible aria-label.
- Speech recognition / Voice control users won't be able to activate the control via a voice command, because of the label / accessible name mismatch, at least with _some_ of the voice control software. See discussion on https://github.com/WordPress/gutenberg/issues/22332

The Web Content Accessibility Guidelines success criterium [Label in Name](https://www.w3.org/TR/WCAG22/#label-in-name) states that the accessible name must _contain_ the visible label and that the best practice is to have the text of the visible label at the start of the name.

Unfortunately, that doesn't work with some voice control software, which makes the WCAG criterium arguably useful.

As such, each time there's a mismatch between visible label and actual accessible name, a new accessibility barrier is introduced for some users. This kind of mismatch should be avoided when possible.

I'd like to reinstate a good general principle:

> [!IMPORTANT]
> When possible, let the visible label of a control be its accessible name. Do not override or expand with screen reader text or aria-label / aria-labelledby.

A good example of _unnecessary_ mismatch is in the Global Styles root menu. Screenshot:

Screenshot 2024-09-06 at 09 57 46

Where the buttons _visible labels_ are:

- Typography
- Colors
- Shadows
- Layout

However, these buttons do have [aria-label attributes](https://github.com/WordPress/gutenberg/blob/1a8aa20d057e6bd623d3ff8178b2d67de2845dcc/packages/edit-site/src/components/global-styles/root-menu.js#L44-L64) that only adds the word 'styles'. As such, their accessible name becomes, respectively:

- Typography styles
- Colors styles
- Shadow styles
- Layout styles

This mismatch prevents users of some voice control software to use a voice command.

In this specific case, the added word 'styles' doesn't add much value. These buttons are already placed in a panel that has a heading 'Styles'. The context appears to be clear enough and the mismatch should be avoided. Keeping things simple is always the best option.

When some meaningful, useful, additional context and information are necessary, a better option would be to place the extra info in the control description rather than unnecessarily altering the accessible name.

### Step-by-step reproduction instructions

- Go to Site Editor > Styles
- Observe the root menu in the Styles panel.
- Observe the visible label of the buttons Typography, Colors, Shadows, Layout, mismatches the accessible name provided by the aria-label attributes.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

- [X] Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

- [X] Yes

### PR DESCRIPTION
**Pull request by afercia**
Fixes https://github.com/WordPress/gutenberg/issues/65112

## What?

Removes most aria-label that:
- Either alter the accessible name thus introducing a mismatch between visible label and actual accessible name
- Or just repeat the visible text.

## Why?

In such cases, the labels are unnecessary or even harmful for accessibility. Always prefer the visible label to determine the accessible name.

## How?

- Removes unnecessary aria-labels.
- Adjust the tests.
- Special case: Adds a visible label to the 'Palette' button. Note: this is the only visual change in this PR.
- Removes an unnecessary HTML `title` attribute.

Screenshot of the Palette butto, before and after:

Screenshot 2024-09-06 at 14 54 06

## Testing Instructions

- Go to the Site Editor > Styles
- Observe the button items in the root menu and in the various submenus have an accessible name determined by the visible label i.e. no `aria-label` attribute.

### CODE REVIEW

**Comment by Mamaduka** [comment]
Happy to help with the code review, but I think this also needs design feedback.

cc @WordPress/gutenberg-design

**Review by richtabor — DISMISSED** [review]
> Special case: Adds a visible label to the 'Palette' button. Note: this is the only visual change in this PR.

This does not following existing UI patterns. I'd encourage not introducing new UI for existing concepts.

**Comment by afercia** [comment]
> > Special case: Adds a visible label to the 'Palette' button. Note: this is the only visual change in this PR.
> 
> This does not following existing UI patterns. I'd encourage not introducing new UI for existing concepts.

The current label made of 'palette icons' of the control is already a unique case that doesn't provide a choesive experience for users. It's not accessible and to me it's not a great design. I see two options:

- Add a label and keep the palette icons, to at least partially respect the original design.
- Otherwise, the only alternative is to entirely remove the palette icons. 

I'd like to be very clear that controls labeled with only icons must be avoided, especially when the icons are not recognizable and can't be mentally associate to the name of the control.

> I'd encourage not introducing new UI for existing concepts.

Any feedback is welcomed but saying 'No' _without providing an alternative solution to a valid issue_ is not acceptable and not welcomed. I think I already pointed out on other issues and PRs that this kind of feedback doesn't help in any way to make WordPress better and doesn't help collaboration in an open source project. Thanks.

**Comment by afercia** [comment]
In the last commit I'm entirely removing the 'palette icons' and only using visible text for the label.

To recap: these icons used as label illustrated in the screenshot below are not OK. This is not the kind of accessible, inclusive, design I'd liek to see in WordPress:

not ok

The control is now labeled with visible text only:

Screenshot 2024-09-12 at 10 54 13

As a consequence, the 'Theme colors randomizer' button that is shown when enabling the 'Color randomizer' Gutenberg experiment is now moved one level deeper into the 'edit palette' panel. Screenshot:

Screenshot 2024-09-12 at 10 55 05

**Comment by afercia** [comment]
One more good reason to move the 'Randomize color' button to the actual Edit palette panel is that the colors icons shown inside the 'Edit palette' button are only [the first five colors provided by the theme](https://github.com/WordPress/gutenberg/blob/9bdfebfa6760fcc9f1d11a185ecb2940d475fcb5/packages/edit-site/src/components/global-styles/palette.js#L74). Depending on the theme palette, clicking 'Randomize color' may not show any visual change in the five colors icons. For example, in Twenty Twenty-Four the first five colors are all white, black or grays and clicking the Randomize button won't show any change in the colors icons.

**Comment by afercia** [comment]
I'd appreciate a new review of this PR. The only visible change is about the 'palette icons'. As mentioned earlier, that design pattern (which as far as I can tell is a _unique_ case in the editor and as such is an ad-hoc, inconsistent, Implementation that should be avoided in the first place is not only not accessible but it's also non-functional as it may not show any change when randomizing colors, depending on the current palette.

I'm not opposed to alternative solutions as long as:
- It's not a unique, ad-hoc, implementation.
- It's accessible.
- It works with any color palette. currently, it doesn't. See my previous comment https://github.com/WordPress/gutenberg/pull/65124#issuecomment-2345944052

Thanks

Cc @WordPress/gutenberg-core @WordPress/gutenberg-design

**Comment by jameskoster** [comment]
Couldn't the label be inline, to at least resemble the format of the items below (which are also edit affordances). I don't love it, but without a broader redesign of these elements I don't see an obvious better solution.

Screenshot 2024-10-14 at 13 19 57

**Comment by afercia** [comment]
@jameskoster thanks for your feedback, that could be an option. However, did you try this UI with Twenty Twenty-Four set as active theme? Did you check that, when randomizing colors, those 5 icons don't change at all? As I mentioned, those icons preview is non-functional with some color palettes. I'm not sure why the editor should keep an UI that:
- it doesn't work well and can't be guaranteed to work well because we can't predict the palette colors.
- it's a unique design pattern in the editor UI, this is the only case I can think of a button that is visually labeled with a non-very intuitive series of colored circles.

I think these colored icons preview should be moved one level deeper into the 'edit palette' panel and ideally not be arbitrarily limited to 5 colors. They should preview the changes for the entire palette.

**Comment by afercia** [comment]
To illustrate visually, with Twenty Twenty-Four active:
- Before: no visible changes when randomizing
- After: Randomize button moved into the edit palette panel so that the randomization shows changes to all colors. Note the first five colors don't (or barely do) change.

| Before  | After |
| ------------- | ------------- |
| before | after |

**Comment by jameskoster** [comment]
The color randomizer is an experimental feature, so I'm not sure we should be adjusting production UI around that. That said I agree about moving the 'Randomize' button to the Edit palette panel. Ideally the 'Edit palette' button should be updated to support random palettes.

Separately I wonder if we should update the swatches in these elements to support multiple colors in a way that doesn't affect the overall footprint:


palette

**Comment by afercia** [comment]
Screenshot 2024-10-14 at 15 53 27

@jameskoster interesting. I wouldn't be opposed to add an icon that represents the concept of 'palette' to the left of the button label. If you can provide one that works well, I'd be glad to add it.

However, I don't think that icon can be a functional preview of the actual palette colors. It can't be arbitrarily limited to 5 colors.

Any objections to move further improvements to a follow-up? I'd like to unblock this PR because it improves many other parts, not just this button.

**Comment by t-hamano** [comment]
I think the purpose of this PR is to resolve the label inconsistency. Can we discuss in another issue/PR whether the location of the color randomizer is appropriate and whether the 5 color icons make sense?

That is, simply add "Edit Palette" to the current button:

image

If the 5 color icons are considered useless, that might mean the palette variation button is useless too.

Below are the palette variation buttons in TT4. Note that only 4 colors are shown here, so the difference is more difficult to see.

image

**Comment by afercia** [comment]
> That is, simply add "Edit Palette" to the current button:

That is what I proposed at the beginning of this PR?

Anyways, I think moving the Randomizer button is an improvement and I don't see good reasons to not do it in this PR. It's a small, self-contained change. I do see way more complex PRs that introduce changes not strictly related to the original scope of the issue to solve being merged without problems. Sometimes, it's inevitable that the original scope of an issue gets bigger and I'd think the opportunity of a change should be evaluated on a case by case basis to avoid serious 'scope creep', but this isn't the case.

**Comment by t-hamano** [comment]
> That is what I proposed at the beginning of this PR?

That's true, but my suggestion is to place it to the right of the color icon, not above it. What do you think about this design? @WordPress/gutenberg-design 

> I do see way more complex PRs that introduce changes not strictly related to the original scope of the issue to solve being merged without problems

While this can certainly happen, I think we should try to narrow it down as much as possible to the problem we're trying to fix. Otherwise, someone looking at a PR for the first time won't know what we actually changed without reading the code and discussion. Also, I think the smaller the PR, the easier it will be to identify the problem later on if we run into a problem.

**Comment by jameskoster** [comment]
> What do you think about this design?

It works for me. 

And agree we should keep the scope of the PR narrow. Moving the randomizer, changing icons, etc. should be discussed separately.

**Comment by ciampo** [comment]
Sounds good to me (and thank you @t-hamano for keeping the scope of this PR narrow).

One last aspect to consider: what do we do if the text doesn't fit in the available space? Truncation or break to a new line? (@jameskoster )

**Comment by afercia** [comment]
> While this can certainly happen, I think we should try to narrow it down as much as possible to the problem we're trying to fix. Otherwise, someone looking at a PR for the first time won't know what we actually changed without reading the code and discussion. Also, I think the smaller the PR, the easier it will be to identify the problem later on if we run into a problem.

Fair enough. I will revert the moving of the randomizer and create another issue and PR.
I generally agree with keeping the scope as small as possible. From now on, I expect to see consistent feedback from you all for all future issues and PRs. Thanks.

**Comment by jameskoster** [comment]
> Truncation or break to a new line?

I'd lean toward the latter.

**Comment by afercia** [comment]
Truncation is not a good content strategy. The FlexItem component allows text wrapping and that's the best option IMHO.

**Comment by afercia** [comment]
Latest commit restores the 'Randomize colors' button, the colors preview, and adds a visible label 'Edit palette'. Screenshot:

Screenshot 2024-10-16 at 13 43 11

The edge case with a longer string and text wrapping is minor because the colors preview will be removed in a follow-up which will also move the Randomize colors to the edit palette panel.

**Comment by jameskoster** [comment]
Would it be better to wrap the text underneath the swatches to avoid situations where many colors create a very narrow space for the text, and the button grows vertically to extreme heights?

**Review by ntsekouras — APPROVED** [review]
Code wise this LGTM, thanks! The only possible small remaining thing is @jameskoster [comment](https://github.com/WordPress/gutenberg/pull/65124#issuecomment-2417024117). Personally I think the current wrapping looks good, but 🤷

**Comment by jameskoster** [comment]
My concern is this:

Screenshot 2024-10-16 at 18 54 24

IMO this works better:

Screenshot 2024-10-16 at 18 56 25

But I acknowledge this is totally an edge case. It doesn't need to hold up the PR :)

**Review by t-hamano — APPROVED** [review]
LGTM!

> Would it be better to wrap the text underneath the swatches to avoid situations where many colors create a very narrow space for the text, and the button grows vertically to extreme heights?

For now, the number of swatches will [not exceed five](https://github.com/WordPress/gutenberg/blob/50aab34e5255a4f7a09f421a24d41211c689e4b4/packages/edit-site/src/components/global-styles/palette.js#L68), so I think there is no need to worry about increasing the number of swatches.

In the future, when we decide whether to delete swatches, change the design of swatches, or increase the swatch limit, we will be able to adjust the text layout.

**Comment by afercia** [comment]
> Would it be better to wrap the text underneath the swatches to avoid situations where many colors create a very narrow space for the text, and the button grows vertically to extreme heights?

@jameskoster @ntsekouras [the colors are limited to five](https://github.com/WordPress/gutenberg/blob/3e2e434c55347584b25273006a03ce82677c3ee3/packages/edit-site/src/components/global-styles/palette.js#L72-L82). Which is, as I said, an arbitrary number that makes the swatches useless with some color palettes. That's how it works right now and this PR isn't changing that. As such, the edge case illustrated in @jameskoster [screenshot above](https://github.com/WordPress/gutenberg/pull/65124#issuecomment-2417530079) can't happen. This button shows only 5 colors.

I created https://github.com/WordPress/gutenberg/issues/66169 to discuss removing the swatches and moving the Randomizer to the edit palette panel.

**Comment by afercia** [comment]
I'd appreciate an approval... Thanks.

**Comment by ntsekouras** [comment]
>I'd appreciate an approval... Thanks.

You have two :)

**Pull request merged**
